traits: Introduce std feature

vks opened the pull request:
This makes it possible to build `traits` without `std`. For this a new
trait `BasicFloat` was introduced, implementing some basic functionality
that works with `core`. Most notably this is lacking functions like
`cos`, `sin`, etc.

`Float` is not available without `std`.

Refs #216.

cuviper commented:
There's a CI failure reported across all versions.  I think this probably needs some direct no-std testing in `ci/test_full.sh` too.

Did you consider just gating `#[cfg(feature = "std")]` on `Float` methods as needed?  I know that makes a lot more places that have to be marked, but I think it's easier to explain.  I don't like the method duplication, and I don't want users to be confused about when to use `Float` or `BasicFloat`.

If you really want to have two separate traits, maybe it would be better to have `Float: BasicFloat` so they are directly related.

cuviper commented:
> maybe it would be better to have `Float: BasicFloat`

Hmm, but that's a breaking change for implementors. :/

vks commented:
> Hmm, but that's a breaking change for implementors. :/

Yes, I considered that, but discarded it for this reason.

I like your suggestion of gating the methods, this would have the advantage that there will be no redundant reimplementation of the methods currently in `BasicFloat` if `std` is available.

vks commented:
I did what you suggested, which coincidentally also fixed the tests.

I added quite a few default implementations to `Float`. They only matter for `no_std` at the moment.

cuviper commented on traits/src/float.rs:
```diff
@@ -402,7 +455,34 @@ pub trait Float
+            }
+        }
+        acc
+    }
```
Maybe use the generic `num_traits::pow()` once `exp` is positive?  I see your implementation is identical, but we don't need to duplicate code.

vks replied:
The problem is that `num_traits::pow` takes a `usize` for the exponent, which might be smaller than `i32` in the future. That is why I duplicated the implementation.

If we made a breaking change, then we could consolidate the two implementations by making `num_traits::pow` use a `u32` for the exponent.

cuviper replied:
Hmm, do 16-bit targets even have these floats?  But OK, not a big deal.

vks replied:
Probably not. I could use `num_traits::pow` and cast, hoping it works. Alternatively we could use a macro to deduplicate the implementation.

cuviper replied:
Or instead of casting, use `FromPrimitive` and `unwrap`.

cuviper commented on traits/src/float.rs:
````diff
@@ -537,13 +625,28 @@ pub trait Float
+    /// assert!(abs_difference < 1e-10);
+    /// ```
+    #[cfg(not(feature = "std"))]
+    fn to_degrees(self) -> Self;
````
Ugh, that's annoying to have a separate declaration, but I don't see a better way, other than excluding it from `no_std` entirely.

cuviper commented on traits/src/float.rs:
```diff
@@ -1097,117 +1287,138 @@ macro_rules! float_impl {
                 // NB: `f32` didn't stabilize this until 1.7
                 // <$T>::to_degrees(self)
-                self * (180. / ::std::$T::consts::PI)
+                self * (180. / ::core::$T::consts::PI)
```
Our baseline is now 1.8 - can we just use the native implementation?

cuviper commented on traits/src/float.rs:
```diff
@@ -1097,117 +1287,138 @@ macro_rules! float_impl {
                 // NB: `f32` didn't stabilize this until 1.7
                 // <$T>::to_radians(self)
-                self * (::std::$T::consts::PI / 180.)
+                self * (::core::$T::consts::PI / 180.)
```
Ditto on using the native impl.

vks commented:
@cuviper I think I addressed all your comments.

cuviper commented on traits/src/float.rs:
```diff
@@ -950,264 +1094,312 @@ macro_rules! float_impl {
-                // NB: `f32` didn't stabilize this until 1.7
-                // <$T>::to_degrees(self)
-                self * (180. / ::std::$T::consts::PI)
+                <$T>::to_degrees(self)
```
Oh, this doesn't work for no_std, because the inherent methods are in std.  So it recurses:

```rust
warning: function cannot return without recurring
    --> src\float.rs:1272:13
     |
1272 |               fn to_degrees(self) -> Self {
     |  _____________^ starting here...
1273 | |                 <$T>::to_degrees(self)
1274 | |             }
     | |_____________^ ...ending here
...
1482 |   float_impl!(f32 integer_decode_f32 classify_f32);
     |   ------------------------------------------------- in this macro invocation
     |
     = note: #[warn(unconditional_recursion)] on by default
```

I guess we have to roll our own, but it's not a big deal since these are trivial.

Can you add `#![deny(unconditional_recursion)]` to the crate?  This seems like a problem we're vulnerable to in general, with so many traits that are supposed to forward calls...

vks replied:
Sorry, I actually wanted to keep the implementation for `no_std` but forgot about it...

Should be fixed in the most recent commit.

cuviper commented:
@homu r+

homu commented:
:pushpin: Commit ba73ba2 has been approved by `cuviper`

<!-- @homu r=cuviper ba73ba2af063b140be8bde0933d9a109dd977535 -->

homu commented:
:hourglass: Testing commit ba73ba2 with merge 8b5d4ac...

homu commented:
:sunny: Test successful - [status](https://travis-ci.org/rust-num/num/builds/241074464?utm_source=github_status&utm_medium=notification)

nox commented:
This is a breaking change and the release should be yanked. Feature-gating things that were previously not feature-gated is a breaking change.
